skyllh/analyses/i3/publicdata_ps/signal_generator.py

```diff
@@ -0,0 +1,563 @@
+        """
+        if gauss is None and box is None:
+            raise ValueError(
+                "Either box or gauss keywords must define the neutrino flare.")
+        if gauss is not None and box is not None:
+            raise ValueError(
+                "Either box or gauss keywords must define the neutrino flare, "
+                "cannot use both.")
+
+        self.box = box
+        self.gauss = gauss
+
+    def generate_signal_events(self, rss, mean, poisson=True):
+        """ same as in PDSignalGenerator, but we assign times here. 
+        """
+        shg_list = self._src_hypo_group_manager.src_hypo_group_list
+
+        tot_n_events = 0
+        signal_events_dict = {}
+
+        for shg in shg_list:
+            # This only works with power-laws for now.
+            # Each source hypo group can have a different power-law
+            gamma = shg.fluxmodel.gamma
+            weights, _ = self.llhratio.dataset_signal_weights([mean, gamma])
+            for (ds_idx, w) in enumerate(weights):
+                w_mean = mean * w
+                if(poisson):
+                    n_events = rss.random.poisson(
+                        float_cast(
+                            w_mean,
+                            '`mean` must be castable to type of float!'
+                        )
+                    )
+                else:
+                    n_events = int_cast(
+                        w_mean,
+                        '`mean` must be castable to type of int!'
+                    )
+                tot_n_events += n_events
+
+                events_ = None
+                for (shg_src_idx, src) in enumerate(shg.source_list):
+                    ds = self._dataset_list[ds_idx]
+                    sig_gen = PDDatasetSignalGenerator(
+                        ds, src.dec, self.effA[ds_idx], self.sm[ds_idx])
+                    if self.effA[ds_idx] is None:
+                        self.effA[ds_idx] = sig_gen.effA
+                    if self.sm[ds_idx] is None:
+                        self.sm[ds_idx] = sig_gen.smearing_matrix
+                    # ToDo: here n_events should be split according to some
+                    # source weight
+                    events_ = sig_gen.generate_signal_events(
+                        rss,
+                        src.dec,
+                        src.ra,
+                        shg.fluxmodel,
+                        n_events,
+                        energy_cut_spline=self.splines[ds_idx],
+                        cut_sindec=self.cut_sindec[ds_idx]
+                    )
+                    if events_ is None:
+                        continue
+
+                    # Assign times for flare. We can also use inverse transform
+                    # sampling instead of the lazy version implemented here
+                    tmp_grl = self.data_list[ds_idx].grl
+                    for event_index in events_.indices:
```

Comment:
Won't fix this now. It will be much easier to implement in the new version, where we could sample random events using cdf of TimePDF which accounts for grl intervals